src/operator/nn/mkldnn/mkldnn_sum.cc

```diff
@@ -48,28 +49,119 @@ void Sum(const mkldnn::memory &arr1, const mkldnn::memory &arr2,
+
+class MKLDNNSumFwd {
+ public:
+  explicit MKLDNNSumFwd(const std::vector<NDArray> &inputs) {
+    _Init(inputs);
+  }
+  ~MKLDNNSumFwd() {}
+  void SetDataHandle(const std::vector<NDArray> &inputs,
+                     const NDArray &output,
+                     const OpReqType &req);
+  void Execute(const NDArray &output, const OpReqType &req);
+
+ private:
+  void _Init(const std::vector<NDArray> &inputs);
+
+ private:
+  std::shared_ptr<mkldnn::sum> fwd;
+  std::vector<std::shared_ptr<mkldnn::memory>> in_data;
+  std::shared_ptr<mkldnn::memory> out;
+  std::shared_ptr<mkldnn::sum::primitive_desc> fwd_pd;
+};
+
+void MKLDNNSumFwd::_Init(const std::vector<NDArray> &inputs) {
+  std::vector<mkldnn::memory> in_mems;
   for (size_t i = 0; i < inputs.size(); i++) {
-    auto in_mem = inputs[i].GetMKLDNNData();
-    in_prims.push_back(*in_mem);
-    in_pds[i] = in_mem->get_primitive_desc();
+    in_mems.push_back(*inputs[i].GetMKLDNNData());
+  }
+  this->in_data.resize(inputs.size());
+  for (size_t i = 0; i < in_mems.size(); i++) {
+    this->in_data[i].reset(new mkldnn::memory(in_mems[i].get_primitive_desc(),
+                                              in_mems[i].get_data_handle()));
+  }
+
+  std::vector<float> scales(inputs.size());
+  std::vector<mkldnn::memory::primitive_desc> in_pds(inputs.size());
+  for (size_t i = 0; i < in_mems.size(); i++) {
+    in_pds[i] = in_mems[i].get_primitive_desc();
     scales[i] = 1;
   }
-  mkldnn::sum::primitive_desc pdesc(scales, in_pds);
+  this->fwd_pd.reset(new mkldnn::sum::primitive_desc(scales, in_pds));
+  this->out.reset(new mkldnn::memory(this->fwd_pd->dst_primitive_desc()));
+
+  std::vector<mkldnn::primitive::at> in_prims;
+  for (size_t i = 0; i < this->in_data.size(); i++) {
+    in_prims.push_back(*(this->in_data[i]));
+  }
 
-  auto out_mem = CreateMKLDNNMem(out_data, pdesc.dst_primitive_desc(), req);
+  this->fwd.reset(new mkldnn::sum(*(this->fwd_pd), in_prims,
+                                  *(this->out)));
+}
+
+void MKLDNNSumFwd::SetDataHandle(const std::vector<NDArray> &inputs,
+                                 const NDArray &output,
+                                 const OpReqType &req) {
+  std::vector<mkldnn::memory> in_mems;
+  for (size_t i = 0; i < inputs.size(); i++) {
+    in_mems.push_back(*inputs[i].GetMKLDNNData());
+    this->in_data[i]->set_data_handle(in_mems[i].get_data_handle());
+  }
+
+  auto out_mem = CreateMKLDNNMem(output, this->fwd_pd->dst_primitive_desc(),
+                                 req);
+  this->out->set_data_handle(out_mem.second->get_data_handle());
+}
+
+void MKLDNNSumFwd::Execute(const NDArray &output, const OpReqType &req) {
   MKLDNNStream *stream = MKLDNNStream::Get();
-  stream->RegisterPrim(mkldnn::sum(pdesc, in_prims, *out_mem.second));
-  CommitOutput(out_data, out_mem);
+  stream->RegisterPrim(*(this->fwd));
+  auto out_mem = CreateMKLDNNMem(output, this->fwd_pd->dst_primitive_desc(),
+                                 req);
```

Comment:
you shouldn't call CreateMKLDNNMem twice. The way you restructure the code makes it difficult to work with the interface I designed.

Reply:
In order to only invoke CreateMKLDNNMem() once and keep a unified SetDataHandle/Execute API interface (pass NDArray as parameters), one way is to define 2 members under 
 MKLDNNSumFwd: "std::shared_ptr<mkldnn::memory> out" and "OutDataOp op" to save created memory and data_op and use it later (maybe a little ugly). Please let me know your suggestions as well if there is better choice.